Tensile/Components/MAC_F32.py

```diff
@@ -80,19 +87,56 @@ def __call__(self, writer, m, innerUnroll):
                     vars["aStr"] = "v[vgprValuA_X{m}_I{iui} + {a}]".format_map(vars)
                     vars["bStr"] = "v[vgprValuB_X{m}_I{iui} + {b}]".format_map(vars)
 
-                    if instruction == "v_fma_f32":
-                        kStr += "v_fma_f32 {cStr}, {aStr}, {bStr}, {cStr}{endLine}".format_map(vars)
+                    instVars = {}
+                    instVars["endLine"] = writer.endLine
+                    instVars["cStr"] = vars["cStr"]
+                    instVars["aStr"] = vars["aStr"]
+                    instVars["bStr"] = vars["bStr"]
+
+                    if dualMacEnable == 1:
+                        if instQ.empty():
+                            instQ.put(instVars)
+                        else:
+                            # pop instruction
+                            prevVars = instQ.queue[0]
+                            if instVars["cStr"] != prevVars["cStr"]:
+                                # make dual fmac
+                                kStr += "v_dual_fmac_f32 {cStr}, {aStr}, {bStr}".format_map(prevVars) + " :: v_dual_fmac_f32 {cStr}, {aStr}, {bStr}{endLine}".format_map(vars)
+                                kStr += priority(writer, 1, "Raise priority while processing macs")
+                                instQ.get()
+                            else:
+                                # push instruction
+                                instQ.put(instVars)
                     else:
-                        kStr += "{instruction} {cStr}, {aStr}, {bStr}{endLine}".format_map(vars)
-
-                    kStr += priority(writer, 1, "Raise priority while processing macs")
+                        if instruction == "v_fma_f32":
+                            kStr += "v_fma_f32 {cStr}, {aStr}, {bStr}, {cStr}{endLine}".format_map(vars)
+                        else:
+                            kStr += "{instruction} {cStr}, {aStr}, {bStr}{endLine}".format_map(vars)
 
                     if macIdx == kernel["PerformanceWaitLocation"]:
+                        kStr += self.popAllInstructions(instruction, instQ, priority, writer)
+                        instQ.queue.clear()
```

Comment:
I think clear() is redundant.
popAllInstructions() ensures the size of instQ is 0.

Reply:
Thanks. The queue object is mutable.